_drafts/implementing-state-machines-in-postgresql.md

```diff
@@ -0,0 +1,301 @@
+our `order_events` table that makes sure all events of a given `order_id`
+are valid and don't lead to an error state. We do this using a simple `plpgsql`
+function that executes our `order_events_fsm` against all existing events
+of the current `order_id` plus the new event. If the final state is `error`,
+we raise an exception which causes the current transaction to roll back. It
+looks like this:
+
+
+```sql
+CREATE FUNCTION order_events_tigger_func() RETURNS trigger
+LANGUAGE plpgsql AS $$
+DECLARE
+  new_state text;
+BEGIN
+  SELECT order_events_fsm(event ORDER BY id)
+  FROM (
+    SELECT id, event FROM order_events WHERE order_id = new.order_id
+    UNION
+    SELECT new.id, new.event
+  ) s
+  INTO new_state;
+  
+  IF new_state = 'error' THEN
+    RAISE EXCEPTION 'invalid event';
+  END IF;
+
+  RETURN new;
+END
+$$;
+
+CREATE TRIGGER order_events_trigger BEFORE INSERT ON order_events
+FOR EACH ROW EXECUTE PROCEDURE order_events_tigger_func();
+```
+
+We can verify that this works by inserting a valid sequence of order events:
+
+```sql
+INSERT INTO order_events (order_id, event) VALUES
+  (1, 'create'),
+  (1, 'pay'),
+  (1, 'ship');
+```
+```
+INSERT 0 3
+```
+
+As well as an invalid sequence of events:
+
+```sql
+INSERT INTO order_events (order_id, event) VALUES
+  (2, 'create'),
+  (2, 'ship');
+```
+```
+psql:orders.sql:95: ERROR:  invalid event
+CONTEXT:  PL/pgSQL function order_events_tigger_func() line 14 at RAISE
+```
+
+As expected, only the first series of inserts made it into our table:
+
+```sql
+SELECT id, order_id, event FROM order_events;
+```
+```
+ id | order_id | event
+----+----------+--------
+  1 |        1 | create
+  2 |        1 | pay
+  3 |        1 | ship
+```
+
+If you're still on the fence about embedding this kind of logic into your db,
+let's have a look how our approach gives us advanced analytical powers as a
+free by-product. Let's consider a new data set of 3 orders:
+
+
+```sql
+TRUNCATE order_events;
+INSERT INTO order_events (order_id, event, time) VALUES
+  (1, 'create', '2017-07-23 00:00:00'),
+  (1, 'pay', '2017-07-23 12:00:00'),
+  (1, 'ship', '2017-07-24 00:00:00'),
+
+  (2, 'create', '2017-07-23 00:00:00'),
+  (2, 'cancel', '2017-07-24 00:00:00'),
+
+  (3, 'create', '2017-07-23 00:00:00'),
+  (3, 'pay', '2017-07-24 00:00:00'),
+  (3, 'cancel', '2017-07-25 00:00:00'),
+  (3, 'refund', '2017-07-26 00:00:00');
+```
+
+Using our `order_events_fsm` as a [window
+function](https://www.postgresql.org/docs/9.6/static/tutorial-window.html), we
+can easily get the state history of a given order:
+
+```sql
+SELECT time, order_events_fsm(event) OVER (ORDER BY id)
+FROM order_events
+WHERE order_id = 3;
+```
+```
+        time         | order_events_fsm
+---------------------+-------------------
+ 2017-07-23 00:00:00 | awaiting_payment
+ 2017-07-24 00:00:00 | awaiting_shipment
+ 2017-07-25 00:00:00 | awaiting_refund
+ 2017-07-26 00:00:00 | canceled
+```
+
+Or perhaps even more interesting, we can use a
```

Comment:
This sentence feels kinda "clunky". I'd change it to something like "Even perhaps more interesting is using a Lateral sub-query [...]"